crates/forge-runner/src/lib.rs

```diff
@@ -0,0 +1,594 @@
+                    matches!(
+                        item,
+                        TestCaseSummary::Passed { .. } | TestCaseSummary::Failed { .. }
+                    )
+                })
+                .count(),
+        )?;
+
+        match final_result {
+            Some(result) => Ok(result.with_runs(runs)),
+            None => panic!("Test should always run at least once"),
+        }
+    })
+}
+
+#[allow(clippy::too_many_arguments)]
+fn run_fuzzing_subtest(
+    args: Vec<Felt252>,
+    case: Arc<TestCase>,
+    runner: Arc<SierraCasmRunner>,
+    runner_config: Arc<RunnerConfig>,
+    runner_params: Arc<RunnerParams>,
+    cancellation_tokens: Arc<CancellationTokens>,
+    cancellation_fuzzing_token: CancellationToken,
+    send: Sender<()>,
+    send_shut_down: Sender<()>,
+) -> JoinHandle<Result<TestCaseSummary>> {
+    task::spawn(async move {
+        tokio::select! {
+            () = cancellation_tokens.error.cancelled() => {
+                // Stop executing all tests because
+                // one of a test returns Err
+                Ok(TestCaseSummary::Skipped {  })
+            },
+            () = cancellation_tokens.exit_first.cancelled() => {
+                // Stop executing all tests because flag --exit-first'
+                // has been set and one test FAIL
+                Ok(TestCaseSummary::Skipped {  })
+            },
+            () = cancellation_fuzzing_token.cancelled() => {
+                // Stop executing all single fuzzing tests
+                // because one of fuzzing test has been FAIL
+                Ok(TestCaseSummary::Skipped {  })
+
+            },
+           result = blocking_run_from_test(
+                args.clone(),
+                case,
+                runner,
+                runner_config.clone(),
+                runner_params.clone(),
+                send.clone(),
+                send_shut_down.clone()
+            ) => {
+                match result? {
+                    Ok(result) => {
+                        if let TestCaseSummary::Failed { .. } = &result {
+                            if runner_config.exit_first {
+                                cancellation_tokens.exit_first.cancel();
+                            }
+                        }
+                        Ok(result)
+                    }
+                    Err(e) => {
+                        cancellation_tokens.error.cancel();
+                        Err(e)
+                    }
+                }
+            },
+        }
+    })
+}
+
+fn function_args<'a>(function: &'a Function, builtins: &[&str]) -> Vec<&'a ConcreteTypeId> {
+    let builtins: Vec<_> = builtins
+        .iter()
+        .map(|builtin| Some(SmolStr::new(builtin)))
+        .collect();
+
+    function
+        .signature
+        .param_types
+        .iter()
+        .filter(|pt| !builtins.contains(&pt.debug_name))
+        .collect()
+}
+
+fn print_test_result(test_result: &TestCaseSummary) {
```

Comment:
Why is it here?

Reply:
I can either copy this code to this file or copy the whole `pretty_printing` to `forge-runner` which doesn't make much sense, because we cannot have a cyclic dependency between `forge-runner` and `forge` crates.

Ideally we'd move it to a separate crate or introduce some different abstraction, but that's for future PRs.

Reply:
It should be at least in a separate file

Reply:
Moved